File: Dalgo Proyecto Uno/ProblemaP1.py
import sys

def solucion(n, j, m, arr):
    # Crear una matriz 3D dp con valores iniciales de infinito
    dp = [[[float('inf')] * (m + 1) for _ in range(n + 1)] for _ in range(n + 1)]
    
    for i in range(n + 1):
        for s in range(j + 1):
            for k in range(m + 1):
                if i == 0:
                    dp[i][s][k] = 0
                elif i == s:
                    dp[i][s][k] = sum(arr[:i])  # Solo suma los primeros i elementos
                elif k == 0:
                    dp[i][s][k] = sumatoria(s, arr)
                elif s > i:
                    dp[i][s][k] = float('inf')
                elif k >= (i - s):
                    dp[i][s][k] = min(dp[i - 1][s][k], dp[i - 1][s - 1][k - (i - s)] + arr[i - 1])
                elif k < (i - s):
                    dp[i][s][k] = dp[i - 1][s][k]
                
    
    return dp[n][j][m]

def sumatoria(j, arr):
    suma = 0
    i = 0
    while i<j :
        suma += arr[i]
        i += 1
    return suma

#print(solucion(23, 11, 19, [127, 103, 1, 23, 81, 43, 61, 153, 181, 47, 7, 3, 27, 91, 43, 57, 21, 1, 73, 13, 13, 1, 31]))
def procesar_casos():
    # Leer la cantidad de casos de prueba
    t = int(sys.stdin.readline().strip())

    for _ in range(t):
        # Leer valores de n, j, m y la lista de enteros
        entrada = list(map(int, sys.stdin.readline().strip().split()))
        n, j, m = entrada[:3]  
        arr = entrada[3:] 
        
        print(solucion(n, j, m, arr))

if __name__ == "__main__":
    procesar_casos()

File: Dalgo Proyecto Uno/ProblemaP1.java
import java.util.*;

public class ProblemaP1 {
    static final int INF = 1000000000; // Valor suficientemente grande

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt(); // Número de casos de prueba
        while (T-- > 0) {
            int n = sc.nextInt();
            int j = sc.nextInt();  // Número de elementos a seleccionar
            int m = sc.nextInt();  // Número máximo de swaps permitidos
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = sc.nextInt();
            }
            System.out.println(solve(n, j, m, arr));
        }
        sc.close();
    }

    // Función que resuelve el problema utilizando DP en dos capas (prev y curr)
    static int solve(int n, int j, int m, int[] arr) {
        // Usamos dos arreglos 2D: prev[s][x] y curr[s][x]
        // s: cantidad de elementos seleccionados (0..n)
        // x: número de swaps usados (0..m)
        int[][] prev = new int[n + 1][m + 1];
        int[][] curr = new int[n + 1][m + 1];

        // Inicializamos prev con INF
        for (int s = 0; s <= n; s++) {
            Arrays.fill(prev[s], INF);
        }
        // Caso base: con 0 elementos, solo s = 0 es posible y su costo es 0
        for (int x = 0; x <= m; x++) {
            prev[0][x] = 0;
        }
        
        // Precomputamos el arreglo de prefijos: prefix[i] = sum(arr[0..i-1])
        int[] prefix = new int[n + 1];
        prefix[0] = 0;
        for (int i = 1; i <= n; i++) {
            prefix[i] = prefix[i - 1] + arr[i - 1];
        }
        
        // Iteramos para i = 1 hasta n (considerando los primeros i elementos)
        for (int i = 1; i <= n; i++) {
            // Inicializamos la capa actual con INF
            for (int s = 0; s <= n; s++) {
                Arrays.fill(curr[s], INF);
            }
            // Caso base: s = 0 siempre es 0
            for (int x = 0; x <= m; x++) {
                curr[0][x] = 0;
            }
            // Para s de 1 hasta i (recordemos que s > i no es posible, y lo dejamos en INF)
            for (int s = 1; s <= i; s++) {
                if (s == i) {
                    // Si se deben seleccionar todos los i elementos, el costo es la suma de ellos.
                    for (int x = 0; x <= m; x++) {
                        curr[s][x] = prefix[i];
                    }
                } else {
                    int cost = i - s; // swaps necesarios para "mover" arr[i-1] a la posición s
                    for (int x = 0; x <= m; x++) {
                        if (x < cost) {
                            // No hay suficientes swaps: no podemos elegir arr[i-1]
                            curr[s][x] = prev[s][x];
                        } else {
                            // Opción 1: No se incluye arr[i-1]
                            int op1 = prev[s][x];
                            // Opción 2: Se incluye arr[i-1] consumiendo 'cost' swaps
                            int op2 = prev[s - 1][x - cost] + arr[i - 1];
                            curr[s][x] = Math.min(op1, op2);
                        }
                    }
                }
            }
            // Copiamos curr a prev para la siguiente iteración
            for (int s = 0; s <= n; s++) {
                System.arraycopy(curr[s], 0, prev[s], 0, m + 1);
            }
        }
        // La respuesta final es dp(n, j, m), almacenado en prev[j][m]
        return prev[j][m];
    }
}
